### DISCUSSION TITLE
Exposure of PhysicsSystem.RestoreState(), .SaveState(), .Update()

### DISCUSSION THREAD

**Discussion by Awarets** [godot-jolt/godot-jolt#623] [General]
I'm making a networked game, and for that it would be useful to be able to rewind and manually fast forward the physics simulation to allow for inaccuracies in clientside input prediction to be rectified.

I took a brief look at [Jolt's API for the PhysicsSystem class](https://jrouwe.github.io/JoltPhysics/class_physics_system.html) and it seems like the .RestoreState(), .SaveState(), and .Update() methods would make this possible.

**Comment by mihe** [comment]
I went ahead and transferred this to a discussion instead. Mostly because I think this seems like the type of feature that should propagate from the Godot side of things in some form, rather than be "sideloaded" through this extension. As an example, there is already an open Godot pull request ([here](https://github.com/godotengine/godot/pull/76462)) for adding the ability to step the physics simulation manually, which I believe can be leveraged to something of the effect you're looking for. It obviously does nothing to deal with the actual saving of state though.

In general I've tried to refrain from creating too much of a disparity between the interface of this extension and the general physics server interface. The biggest reason being that things exposed from a GDExtension will only be (properly) available through GDScript and other dynamic languages, and not statically compiled languages like C#, C++ or Rust. You would need to somehow distribute separate bindings for each of those languages for it to be available there.

But even when disregarding all that, this would likely involve more work than just exposing Jolt's `StateRecorder` stuff, since that on its own won't do much good with regards to saving and restoring the state that's kept within the extension itself, which is plenty. I'm also not sure how you'd go about dealing with the adding/removal of bodies inbetween two states, since that could potentially (I guess) involve needing to reconstruct or destroy the associated scene nodes as well.

**Reply by warent** [reply]
Note: here's the related Godot proposal https://github.com/godotengine/godot-proposals/issues/2821

**Comment by TypicalFooBar** [comment]
Oh, this is what I think I want, as well. I am trying to have my game server run the physics simulation as quickly as possible and then give my clients the state from each step for them to play/rewind/etc. With GodotPhysics, I've only been able to wait the real amount of time that the physics simulation would take and I thought I'd be able to do this with Jolt. Perhaps having this available in godot-jolt would solve this particular scenario! :)

**Comment by albertok** [comment]
It's been over a year now and Godot shows no sign of hiring a physics developer. In the meantime what would be required to add the functonality?

I noticed Godot Rapier went ahead and exposed the calls with a simple api. 
https://godot.rapier.rs/docs/documentation/determinism#manual-stepping

I gave it a go and it works well however there's a few other features I need that are missing/underperforming currently.

**Comment by FabriceCastel** [comment]
In addition to what @albertok mentioned, given Jolt is now more officially integrated into Godot and given this feature request has been outstanding since at least 2019 (with a mostly working PR that's sat unreviewed due to lack of ownership on the physics side of the Godot engine), I think there's good reason to believe exposing this API here first, along with the fact that Godot Rapier also supports them would help push the effort forward upstream.

At this point it seems like waiting for the change to happen in the Godot repo _first_ is going to leave us stuck at a standstill :/

**Comment by mihe** [comment]
As mentioned in the `README.md` of this repository, this extension has entered into maintenance since the integration into Godot itself. I have no intention (nor the time) to do anything but bug fixes to it.
